Database Flags, Engine improvements

naxolotl opened the pull request:
- catch exceptions in WPF initializer (like missing dll)
 - added flags for db entries (so broken files are not rescanned every time) 
 - - also a flag for future use, for the user to manually exclude files (not implemented in UI yet)
 - - extended csv export, option to show only excluded
 - improved dupechecking speed ~50%
 - display-thumbnail generation in separate step (not needed in console), and can run in background for GUI
 - - linux version needs to be adapted

0x90d commented:
Thanks for your work, I'll review it tomorrow likely. Can you tell me what needs to be adapted in linux version? I.e. does any of these change break something in the linux version? (haven't looked through the changes yet)

naxolotl commented:
ScanEngine now does not fill Thumbnail in DuplicateItem when scanning, but there is another Task that can be run: PopulateDuplicateThumbnails() which will do that. So the user can already see the list while the thumbnails are loading.  Or the console version does not need to do it at all. Basically the changes I made in MainWindowVM need to be adapted to the linux version.

Database: the IsImage bool was stored as a uint32 (0: false, 1: true). I replaced it with a Bitfield (again backed by a 32bit uint), so it's 100% backwards compatible. I'm not sure how you want to implement manual exclusion, but one way would be with a flag like I added (but is never set at the moment).

Language: do you use a tool to translate the properties files?

In the future, it might be nice to have a window where the user can browse the database, remove or rescan files where ffmpeg or ffprobe failed, and un-exclude files.

0x90d commented on DuplicateFinderEngine/Data/VideoFileEntry.cs:
```diff
@@ -26,6 +43,11 @@ public VideoFileEntry(string file)
 		[ProtoMember(5)]
-		public readonly bool IsImage;
-    }
+		public EntryFlags Flags;
```
Can this break existing databases when you changed the type by preserving the protonmember field number?

0x90d commented on DuplicateFinderEngine/ScanEngine.cs:
```diff
@@ -64,12 +63,23 @@ public class ScanEngine {
 		}
-		public async void ExportDatabaseVideosToCSV() {
-			await Task.Run(() => DatabaseHelper.ExportDatabaseVideosToCSV(DatabaseFileList));
+		public async void ExportDatabaseVideosToCSV(bool onlyVideos, bool onlyFlagged) {
```
Wouldn't it be better using the flag as paramater you created in VideoFileEntry.cs and then just provide the flags to include / exclude?

naxolotl replied:
i figured there are many ways we might want to use the flags the in the future, right now the expression is: 
(f == ManuallyExcluded || f== AllErrors) && f == IsImage
chances are, if the next version needs more parantheses, a generic solution might need to be changed again

0x90d commented on DuplicateFinderEngine/ScanEngine.cs:
```diff
@@ -117,151 +126,154 @@ public class ScanEngine {
+						return;
+					}
+
+					for (var n = i + 1; n < dupeScanList.Count; n++) {
```
Using i + 1 if the current processing file is near the end, it will only be scanned against a handful of items. You would miss lots of potential dups as the file is only checked against a handful of items then. Did I miss something?
